### PR TITLE
Fix issue with data leaking inside partial `@defer` or `@stream` boundaries during intermediate results

### LINKED ISSUE #13330: Incorrect `dataState`/`data` returned for incremental partial data written to cache with `returnPartialData: false`
### Issue Description

When partial data is written to the cache and `returnPartialData` is either `false` or not set, incremental chunks may still return partial data in a response with a `dataState` of `streaming`.

This is primarily due to these lines in `QueryInfo`:

```ts
const diffOptions = {
  query,
  variables,
  returnPartialData: true,
  optimistic: true,
};

const lastDiff =
  skipCache ? undefined : this.cache.diff<TData>(diffOptions);

let result = this.maybeHandleIncrementalResult(
  lastDiff?.result,
  incoming,
  query
);
```

We unconditionally pull partial data from the cache and apply it with the incremental result which mixes the partial cache data into the final result. See the reproduction in the test case below which demonstrates the issue. What we should get is a result back where the only hole in the data is at the `@defer` boundary and ensure partial data inside that boundary is omitted.

> [!NOTE]
> I (Jerel) plan to implement this fix as it requires code from another pull request and with a specific implementation that will fix this as the cache level. Please do not attempt to fix this yourself.

### Link to Reproduction

n/a

### Reproduction Steps

The following test demonstrates the issue

```ts
test('returns non-deferred cached data with a "cache-first" fetch policy and returnPartialData: false', async () => {
  const query = gql`
    query {
      greeting {
        message
        ... on Greeting @defer {
          recipient {
            name
            email
          }
        }
      }
    }
  `;

  const { httpLink, enqueueInitialChunk, enqueueSubsequentChunk } =
    mockDeferStreamGraphQL17Alpha9();
  const cache = new InMemoryCache();

  // We are intentionally writing partial data to the cache. Suppress console
  // warnings to avoid unnecessary noise in the test.
  {
    using _consoleSpy = spyOnConsole("error");
    cache.writeQuery({
      query,
      data: {
        greeting: {
          __typename: "Greeting",
          recipient: {
            email: "test@example.com",
          },
        },
      },
    });
  }

  const client = new ApolloClient({
    cache,
    link: httpLink,
    incrementalHandler: new GraphQL17Alpha9Handler(),
  });

  const stream = new ObservableStream(client.watchQuery({ query }));

  await expect(stream).toEmitTypedValue({
    data: undefined,
    dataState: "empty",
    loading: true,
    networkStatus: NetworkStatus.loading,
    partial: true,
  });

  enqueueInitialChunk({
    data: { greeting: { message: "Hello world", __typename: "Greeting" } },
    pending: [{ id: "0", path: ["greeting"] }],
    hasNext: true,
  });

  await expect(stream).toEmitTypedValue({
    data: markAsStreaming({
      greeting: {
        __typename: "Greeting",
        message: "Hello world",
      },
    }),
    dataState: "streaming",
    loading: true,
    networkStatus: NetworkStatus.streaming,
    partial: true,
  });

  enqueueSubsequentChunk({
    incremental: [
      {
        data: {
          __typename: "Greeting",
          recipient: { name: "Alice", __typename: "Person" },
        },
        id: "0",
      },
    ],
    completed: [{ id: "0" }],
    hasNext: false,
  });

  await expect(stream).toEmitTypedValue({
    data: {
      greeting: {
        __typename: "Greeting",
        message: "Hello world",
        recipient: { __typename: "Person", name: "Alice" },
      },
    },
    dataState: "complete",
    loading: false,
    networkStatus: NetworkStatus.ready,
    partial: false,
  });

  await expect(stream).not.toEmitAnything();
});
```

### `@apollo/client` version

4.x

### PR DESCRIPTION
**Pull request by jerelmiller**
Fixes #13330

The big change in this PR is making `InMemoryCache` incremental-aware for cache reads (though hidden behind an internal symbol that enables it). `cache.diff` is too naive and classifies in-progress queries with `@defer` boundaries as partial when `@defer` data hasn't streamed in. This is a problem for how `QueryInfo` rereads data after a cache write because it won't apply that data to the returned result unless its fully complete. Field values transformed by cache read functions were therefore not applied to intermediate results in a `@defer` query.

This issue became glaringly apparent with the introduction of custom scalars in 4.3 (currently unreleased) where we need to apply the parsed field values stored in the cache to intermediate chunks. Without that, the user might encounter runtime crashes when expecting the value to be the parsed type.

#13324 was a first attempt at solving this issue across many cases where read functions/custom scalars weren't applied, but it did not fix the issue where partial data inside a `@defer` or `@stream` boundary leaked in intermediate results. It relied on the `missing` tree to reapply field values from the reread `cache.diff`, but did not prune partial boundaries.

`readFromStore` can now properly handle partial boundaries when provided with `returnPartialData: true` (though only when the hidden flag is provided to ensure backwards compatibility until v5). The complex scenarios where partial data can show up inside defer boundaries (especially with overlapping selections in `@defer` boundaries) is now properly handled during a cache read.

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented incomplete cached data from leaking into `network-only` incremental query results.
  * Improved handling of partial `@defer` content and incomplete `@stream` array items.
  * Corrected cache result reuse when registered fragments resolve differently across queries.

* **New Features**
  * Added richer incremental result states: complete, streaming, partial, and empty.
  * Improved tracking of pending deferred and streamed results.
  * Expanded cache memory metrics for incremental processing.

### CODE REVIEW

File: src/cache/inmemory/__tests__/cache.diff/incremental.test.ts
Comment (jerelmiller) [inline src/cache/inmemory/__tests__/cache.diff/incremental.test.ts:1]: Holy tests batman! This is where most of the changes in this PR live. Apologies ahead of time for this being so large. There were MANY edge cases that we never handled properly with `@defer` and `@stream` up to this point and this test suite documents many of those cases. This was especially apparent with partial data when used with `@defer` or `@stream` since `cache.diff` was naive and reported all in-progress `@defer` results as partial.

Do not feel like you have to review every line. Much of the test setup is the same for each test. Things to pay attention to are: the `@defer`/`@stream` boundaries, whether partial data is written inside those boundaries, and whether we ask for partial data back. These tests are a multitude of variations on those scenarios, especially around overlapping fields, sibling `@defer` boundaries, and `@defer`/`@stream` boundaries nested inside each other.

Thanks ahead of time for looking through this 🙂

File: src/cache/inmemory/readFromStore.ts [hunk 1]
```diff
@@ -234,20 +425,47 @@ export class StoreReader {
...
 
-    return {
-      result:
-        complete ? result
-        : returnPartialData ?
-          Object.keys(result).length === 0 ?
-            null
-          : result
-        : null,
+    if (
```
Comment (jerelmiller) [inline src/cache/inmemory/readFromStore.ts:421]: Note: Many of these changes will be simplified when we get to v5 and can enforce caches to be incremental-aware. To maintain backwards compatibility, we can shift data states around. v5 will also require cache reads to return a `dataState` which will collapse this further.

File: src/core/QueryInfo.ts [hunk 2]
```diff
@@ -267,6 +273,7 @@ export class QueryInfo<
       ...incrementalResult,
       dataState:
         incrementalResult.data == null ? "empty"
+          // TODO: This is too naive. For stream arrays, this might be complete
```
Comment (jerelmiller) [inline src/core/QueryInfo.ts]: I need to re-evaluate if this is still true. I'll make sure to remove once I verify if this is ok.

File: src/core/QueryInfo.ts [hunk 3]
```diff
@@ -711,189 +764,3 @@ function shouldWriteResult<T>(
   }
   return writeWithErrors;
 }
-
-function isStreamingPartial(
```
Comment (jerelmiller) [inline src/core/QueryInfo.ts:715]: Much of this is now folded into `readFromStore` which does a prune pass similar to this, but handles many more edge cases that this did/could not.

File: src/utilities/internal/getFragmentFromSelection.ts [hunk 4]
```diff
@@ -24,7 +24,7 @@ export function getFragmentFromSelection(
       }
       const fragment = fragmentMap && fragmentMap[fragmentName];
       invariant(fragment, `No fragment named %s`, fragmentName);
-      return fragment || null;
+      return fragment;
```
Comment (jerelmiller) [inline src/utilities/internal/getFragmentFromSelection.ts:27]: Spotted this while working on this PR. `invariant` checks that this is a non-nullish value so no need to fallback to `null` here.

File: src/core/QueryInfo.ts (hunk 2 above)
Comment (jerelmiller) [reply src/core/QueryInfo.ts]: Re-evaluated and found only one case where it actually was too naive. Fixed in [492337d](https://github.com/apollographql/apollo-client/pull/13347/commits/492337d77d245633530a4f27bcc5aaf24477448d)

File: src/cache/inmemory/readFromStore.ts [hunk 5]
```diff
@@ -189,19 +248,78 @@ export class StoreReader {
...
+          cacheSizes["inMemoryCache.prunePartialBoundaries"] ||
+          defaultCacheSizes["inMemoryCache.prunePartialBoundaries"],
+        makeCacheKey: ({ boundaries, context, selectionSet }) => {
+          if (supportsResultCaching(context.store)) {
+            return this.keyMaker.lookup(
+              selectionSet,
+              boundaries,
+              context.streamInfo,
+              context.deferInfo
+            );
```
Comment (DaleSeo) [inline src/cache/inmemory/readFromStore.ts:265]: The pruned value depends on `path`, but this key omits it?

File: src/cache/inmemory/readFromStore.ts [hunk 6]
```diff
@@ -486,9 +805,183 @@ export class StoreReader {
...
+      // dropped or not.
+
+      const fragment = getFragmentFromSelection(selection, lookupFragment);
+
+      if (
+        fragment &&
+        policies.fragmentMatches(fragment, data.__typename) &&
+        !boundaries.has(selection) &&
+        !context.deferInfo?.peekArray(path)
+      ) {
```
Comment (DaleSeo) [inline src/cache/inmemory/readFromStore.ts:909]: One pending boundary suppresses every fragment at the same response path. How can we prune only the pending boundary?

**Review by DaleSeo — APPROVED** [review]
Just questions.

File: src/cache/inmemory/readFromStore.ts (hunk 5 above)
Comment (jerelmiller) [reply src/cache/inmemory/readFromStore.ts:265]: Yes, we don't really need it due to the objects we use here, but it would actually be problematic as well.

`path` is a new array object each time the `prune*` methods are called. Since it doesn't have a stable identity, it would force a cache miss each time, which defeats the point of caching this function to begin with. Even if we were to stabilize the path identity, we don't really need it because:

- The `selectionSet` is naturally a different object at each path level
- We get a new `boundaries` object anytime `executeSelectionSet` recomputes the object (we could use `data` here instead of `boundaries`, but I chose `boundaries` because it is typically a smaller object)

The only thing `selectionSet` doesn't contain is array indexes, but that is why we have a separate `prunePartialStreamArray` that operates on the whole array, that way we can cache on its `selectionSet` just the same.

Does that make sense?

File: src/cache/inmemory/readFromStore.ts (hunk 6 above)
Comment (jerelmiller) [reply src/cache/inmemory/readFromStore.ts:909]: Not necessarily and in fact we have a ton of different tests in `cache.diff/incremental` that check for overlapping fields in a ton of different ways such as: overlapping sibling `@defer` boundaries with overlapping selections (but different child fields), overlapping deferred/non-deferred fields for the same object, defer/non-deferred fragments at the same level, etc.

As long as a fragment is a sibling of the `@defer` fragment we're pruning, it will still add its selection set to the `workSet` and iterate its selections.

Is there another case not covered by the cache.diff/incremental tests that you're thinking of?

**Review by DaleSeo — APPROVED** [review]
